pkg/apis/serving/v1alpha1/kfservice_types.go

```diff
@@ -34,8 +34,9 @@ var TFExampleKFService = &KFService{
 // KFServiceSpec defines the desired state of KFService
 type KFServiceSpec struct {
 	Default ModelSpec `json:"default"`
-	// Optional Canary definition
-	Canary *CanarySpec `json:"canary,omitempty"`
+	// Canary defines an alternate configuration to route a percentage of traffic.
+	Canary               *ModelSpec `json:"canary,omitempty"`
+	CanaryTrafficPercent int        `json:"canaryTrafficPercent"`
```

Comment:
Since this is moved up, we need to change this to be optional.

Reply:
I thought about this but ended up going my way (option 1) instead. What do you think?

### Option 1
* CanaryTrafficPercent is an int
* defaults to zero, omitted from etcd if not provided.
* No additional validation or rules necessary
* Route code will only use CanaryTrafficPercent if canary is provided
* Pinned Rollout (i.e. percent=0) can be implicit or explicit

### Option 2
* CanaryTrafficPercent is an int*
* defaults to nil, omitted from etcd if not provided.
* validation necessary
* We will fail if canarytrafficpercent provided and not canary itself
* We will fail if canary provided and not canarytrafficpercent
* Pinned Rollout (i.e. percent=0) must be explicit

Reply:
+1 on option 1 as clean and simple. Might still be worth failing validation if canarytrafficPercent>0 and no canary?

Reply:
I think the only issue with option 1 is that if user is only using default, they still needs to have `CanaryTrafficPercent` on spec which may seem a bit weird.

Reply:
~I don't think this is true. It's omitted by default, so they shouldn't see the field.~

Just tested this out. You're right and it's super awkward.

```
apiVersion: serving.kubeflow.org/v1alpha1
  kind: KFService
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"serving.kubeflow.org/v1alpha1","kind":"KFService","metadata":{"annotations":{},"name":"my-model","namespace":"default"},"spec":{"default":{"tensorflow":{"modelUri":"gs://mybucket/mymodel-2"}}}}
    creationTimestamp: 2019-05-19T19:30:57Z
    generation: 1
    name: my-model
    namespace: default
    resourceVersion: "6600042"
    selfLink: /apis/serving.kubeflow.org/v1alpha1/namespaces/default/kfservices/my-model
    uid: a037655e-7a6c-11e9-b0ad-42010a80006a
  spec:
    canaryTrafficPercent: 0
    default:
      tensorflow:
        modelUri: gs://mybucket/mymodel-2
        resources:
          requests:
            cpu: "1"
            memory: 2Gi
        runtimeVersion: 1.13.0
  status:
    canary: {}
    conditions: {}
    default:
      name: my-model-default-zhtl6
    uri: {}
```

Reply:
Just kidding this works with `omitempty`

Going to leave as is, is that okay?

Reply:
Mostly lgtm, but I think it is better to have the validation as @cliveseldon mentioned to fail the validation when `canarytrafficPercent>0` and no canary.

Reply:
Sounds good. I'll do this.